### DISCUSSION TITLE
Licensing C# Scripts for Tabular Editor 3 While Keeping Code Secure

### DISCUSSION THREAD

**Discussion by Rickmaurinus** [TabularEditor/TabularEditor3#1401] [Ideas]
Hi, 

I’ve been working extensively on a great number advanced C# scripts for Tabular Editor 3 that automate complex tasks or dull tasks in Power BI. I’d like to make these scripts available to users on a licensing basis, but I want to ensure the underlying code remains protected and isn’t exposed to the end user.

Does Tabular Editor 3 provide any mechanisms or best practices for securely distributing scripts in a way that allows users to execute them but prevents them from accessing or modifying the code? For example, is there a way to compile scripts or distribute them as an extension?

Looking forward to hear if anyone has tackled this kind of scenario or if there are recommendations for implementing such a distribution model.

Thanks in advance for your input!

Best regards,
Rick de Groot

**Comment by mlonsk** [comment]
Hi Rick (@Rickmaurinus)

To my knowledge, there is currently no way to distribute C# scripts without giving the end-user access to the code.
The technical issue is that Macros can contain any type of code, including malicious code. Therefore, it is the responsibility of the end-user only to run scripts that they are certain are safe, which means they need to be able to see the code they execute. 

We have internally discussed having a Script or Macro store inside Tabular Editor 3, allowing the type of functionality you describe. Configuring it securely will require a significant amount of work from our end, and we do not currently have a timeline for it. You are the first that we know of to suggest this, but if there is demand for such a store both from end-users and script makers for such a feature, that does impact how we prioritize our development. 

I am transferring this discussion into the ideas section as it feels like a feature request to me.

**Reply by Rickmaurinus** [reply]
Hi Morten,

Thanks for your response. I can see the potential security issues, but these scripts took a good amount of time to create. We're not ready to share these publicly at this point. I'll think about this a little more to see if we can come up with a way.

Thanks again for your help!

Best regards,
Rick
